dagql/idtui/streaming_output.go

```diff
@@ -0,0 +1,185 @@
+package idtui
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/dagger/dagger/telemetry/sdklog"
+	"github.com/muesli/termenv"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type streamingExporter struct {
+	spanNames sync.Map
+
+	// incr idx for spanID
+	idx                 int64
+	batchPrinterGetters sync.Map
+	batchPrinterWg      sync.WaitGroup
+
+	output   *termenv.Output
+	outputMu sync.RWMutex
+
+	frameTicker *time.Ticker
+
+	done     chan struct{}
+	doneOnce sync.Once
+}
+
+func newStreamingExporter() *streamingExporter {
+	return &streamingExporter{
+		output:      termenv.NewOutput(os.Stdout),
+		done:        make(chan struct{}),
+		frameTicker: time.NewTicker(50 * time.Millisecond),
+	}
+}
+
+func (t *streamingExporter) Shutdown(ctx context.Context) error {
+	t.doneOnce.Do(func() {
+		t.frameTicker.Stop()
+		close(t.done)
+	})
+
+	// wait all printers outputs
+	t.batchPrinterWg.Wait()
+	return nil
+}
+
+func (t *streamingExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
+	for _, span := range spans {
+		t.spanNames.Store(span.SpanContext().SpanID(), span.Name())
+	}
+	return nil
+}
+
+func (t *streamingExporter) ExportLogs(ctx context.Context, logs []*sdklog.LogData) error {
+	for _, log := range logs {
+		t.export(log)
+	}
+	return nil
+}
+
+func (t *streamingExporter) getSpanName(spanID trace.SpanID) string {
```

Comment:
the span name is not always get cause otel emit the spans until span.End() called.
may consider set span name in log attrs too.